Handle mkldnn broadcasting in mkldnn fuser

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52614 Avoid dispatch overhead in call to mkldnn convolution
* #52613 Include max pool in fusion groups
* #52513 Extend subgraph utils to cover merging a node following a subgraph
* #52512 Remove unused subgraph vmap api
* #51924 Only functionalize ops which we want to include in mkldnn group
* #51923 Add a filter to remove mutation
* #51922 Dont error on 0-dim in convolution
* #51921 Add 0-dim support for binary mkldnn ops
* **#51736 Handle mkldnn broadcasting in mkldnn fuser**
* #51600 Add MKLDNN fuser
* #52786 [NNC] Make NNC sanitize input names
* #51484 Add support for linear in mkldnn fusion
* #51483 Move conv to mkldnn

Differential Revision: [D26696694](https://our.internmc.facebook.com/intern/diff/D26696694)

Commit:
Handle mkldnn broadcasting in mkldnn fuser

[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:
```diff
@@ -300,9 +355,9 @@ class MKLDNNSubgraphSlicer {
-    bool supported_kind = k == prim::MKLDNNGroup || k == aten::to_mkldnn ||
-        k == prim::ConstantMKLDNNTensor;
-    if (supported_kind) {
+    bool supported = k == prim::MKLDNNGroup ||
```
nitpick: if that "if" statement is the only use of `supported` maybe we should just inline it?

Krovatkin commented on torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:
```diff
@@ -34,6 +34,41 @@ using Tensor = at::Tensor;
+      if (a_size.equals(out_size)) {
+        push(stack, a);
+      } else {
+        push(stack, a.to_dense().expand(out_size).to_mkldnn());
```
does this do physical broadcasting?

eellison replied:
yes

Krovatkin commented on test/jit/test_freezing.py:
```diff
@@ -1627,6 +1627,38 @@ def test_unsupported(module, preserved_attrs=None):
+
+            # for good measure, check that broadcasting does not work without this op
+            # so we can remove the op if it ever gets supported
+            with self.assertRaisesRegex(RuntimeError, ""):
```
👍

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "Handle mkldnn broadcasting in mkldnn fuser"




[ghstack-poisoned]

Commit:
Update on "Handle mkldnn broadcasting in mkldnn fuser"




[ghstack-poisoned]

Commit:
Update on "Handle mkldnn broadcasting in mkldnn fuser"




[ghstack-poisoned]

Commit:
Update on "Handle mkldnn broadcasting in mkldnn fuser"




[ghstack-poisoned]

Commit:
Update on "Handle mkldnn broadcasting in mkldnn fuser"




[ghstack-poisoned]

Commit:
Update on "Handle mkldnn broadcasting in mkldnn fuser"




[ghstack-poisoned]

Commit:
Update on "Handle mkldnn broadcasting in mkldnn fuser"




[ghstack-poisoned]

Commit:
Update on "Handle mkldnn broadcasting in mkldnn fuser"




[ghstack-poisoned]

Commit:
Update on "Handle mkldnn broadcasting in mkldnn fuser"




[ghstack-poisoned]

Commit:
Update on "Handle mkldnn broadcasting in mkldnn fuser"


Differential Revision: [D26696694](https://our.internmc.facebook.com/intern/diff/D26696694)

[ghstack-poisoned]
